internal/database/pebble.go

```diff
@@ -152,3 +153,42 @@ func (p *PebbleDB) NewPrefixIterator(prefix []byte) Iterator {
 		p.db.NewIter(prefixIterOptions),
 	}
 }
+
+func (p *PebbleDB) Checkpoint() error {
+	const snapshotDestination = "snapshot"
+	finalDest := filepath.Clean(filepath.Join(p.path, ".."))
+	finalDest = filepath.Join(finalDest, snapshotDestination)
+
+	exists := true
+	_, err := os.Stat(finalDest)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("getting %s stat: %w", finalDest, err)
+		} else {
+			exists = false
+		}
+	}
+
+	// if dir exists then rename it (allowing the checkpoint to be created)
+	// and then remove the _old concurrently
+	if exists {
+		pathToRemove := finalDest + "_old"
+		err = os.Rename(finalDest, pathToRemove)
+		if err != nil {
+			return fmt.Errorf("while renaming: %w", err)
+		}
+
+		go func() {
+			err := os.RemoveAll(pathToRemove)
+			if err != nil {
+				logger.Errorf("failing to remove %s: %w", err)
+			}
+		}()
+	}
+
+	err = p.db.Checkpoint(finalDest)
+	if err != nil {
```

Comment:
If I'm not wrong we can lose snapshots if we delete the "snapshot_old" folder before check that the new snapshot was created.

I'd suggest to execute the deletion only if the new checkpoint was created successful, and if it fails we can rename the "_old" folder back to the previous name to keep being in a consistent state.

What do you think?

Reply:
Agreed, changed to address your suggestion!